crypto/comp/c_zstd.c

```diff
@@ -0,0 +1,830 @@
+/*
+ * Copyright 1998-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ * Uses zstd compression library from https://github.com/facebook/zstd
+ * Requires version 1.4.x (latest as of this writing is 1.4.5)
+ * Using custom free functions require static linking, so that is disabled when
+ * using the shared library.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <openssl/objects.h>
+#include "internal/comp.h"
+#include <openssl/err.h>
+#include "crypto/cryptlib.h"
+#include "internal/bio.h"
+#include "internal/thread_once.h"
+#include "comp_local.h"
+
+COMP_METHOD *COMP_zstd(void);
+
+static COMP_METHOD zstd_method_nozstd = {
+    NID_undef,
+    "(undef)",
+    NULL,
+    NULL,
+    NULL,
+    NULL,
+};
+
+#ifdef OPENSSL_NO_ZSTD
+# undef ZSTD_SHARED
+#else
+
+# ifndef ZSTD_SHARED
+#  define ZSTD_STATIC_LINKING_ONLY
+# endif
+# include <zstd.h>
+
+/* Note: There is also a linux zstd.h file in the kernel source */
+# ifndef ZSTD_H_235446
+#  error Wrong (i.e. linux) zstd.h included.
+# endif
+
+# if ZSTD_VERSION_MAJOR != 1 && ZSTD_VERSION_MINOR != 4
+#  error Expecting version 1.4.x of ZSTD
```

Comment:
So, this PR has been around for so long, that zstd is now on 1.5.2/1.4.10. A bit concerned they are keeping both 1.4.x and 1.5.x alive. Ubuntu 20.04 has 1.4.4, Ubuntu 22.04 has 1.4.8, soI haven't encountered 1.5. I will try out 1.5.2 to see if it still works. As far as I can tell, experimental APIs have been moved to stable. If we want to continue to support 1.4, then I'll have to use the older APIs.

Reply:
I have no strong opinion on what the minimum level we support is. However, one thing we do know is that versions will update over time. I would *assume* that any minor update should be backwards compatible and that major versions would not be. We don't want to (say) lock ourselves into 1.5, if 1.6 will be released next year. So it wouldn't be unreasonable for us to check (for example) `ZSTD_VERSION_MAJOR != 1 || ZSTD_VERSION_MINOR < 5`, and assume we are ok otherwise.

Reply:
Since 1.4 is still being developed, and is in current LTS distributions (e.g. Ubuntu 20 & 22), I think we'd want to keep 1.4 in the mix.